File: ports/pyagentic/pyagentic/web.py
"""Web toolkit (Tier 4) — a robots-aware fetch + text/link extraction tool, built on the
Python standard library (``urllib`` + ``html.parser`` + ``urllib.robotparser``) so it has
**no extra dependencies**. Ships as an opt-in tool, default off: call
``register_web_tools(registry)`` to expose ``web_fetch`` / ``web_links`` to an agent.
Mirrors the Flink ``web/`` toolkit (Jsoup + crawler-commons robots) at a portable subset.
"""

from __future__ import annotations

import urllib.error
import urllib.request
import urllib.robotparser
from html.parser import HTMLParser
from typing import Any, Dict, List, Optional
from urllib.parse import urldefrag, urljoin, urlparse

_DEFAULT_UA = "jagentic-webfetch/0.1 (+https://github.com/jagentic)"
_SKIP_TAGS = {"script", "style", "noscript", "template"}


class _Extractor(HTMLParser):
    """Collects visible text and absolute links from an HTML document."""

    def __init__(self, base_url: str) -> None:
        super().__init__(convert_charrefs=True)
        self._base = base_url
        self._skip_depth = 0
        self._chunks: List[str] = []
        self._title: List[str] = []
        self._in_title = False
        self.links: List[str] = []
        self._seen_links: set[str] = set()

    def handle_starttag(self, tag: str, attrs: List[tuple]) -> None:
        if tag in _SKIP_TAGS:
            self._skip_depth += 1
        if tag == "title":
            self._in_title = True
        if tag == "a":
            for k, v in attrs:
                if k == "href" and v:
                    absolute = urldefrag(urljoin(self._base, v.strip()))[0]
                    scheme = urlparse(absolute).scheme
                    if scheme in ("http", "https") and absolute not in self._seen_links:
                        self._seen_links.add(absolute)
                        self.links.append(absolute)

    def handle_endtag(self, tag: str) -> None:
        if tag in _SKIP_TAGS and self._skip_depth > 0:
            self._skip_depth -= 1
        if tag == "title":
            self._in_title = False

    def handle_data(self, data: str) -> None:
        if self._skip_depth > 0:
            return
        text = data.strip()
        if not text:
            return
        if self._in_title:
            self._title.append(text)
        else:
            self._chunks.append(text)

    @property
    def title(self) -> str:
        return " ".join(self._title).strip()

    @property
    def text(self) -> str:
        # collapse runs of whitespace, keep paragraph-ish separation
        return "\n".join(self._chunks)


def _robots_allows(url: str, user_agent: str, timeout: float) -> bool:
    """Check the site's robots.txt; fail-open if it can't be fetched (the common
    convention), fail-closed only on an explicit Disallow match."""
    parts = urlparse(url)
    robots_url = f"{parts.scheme}://{parts.netloc}/robots.txt"
    rp = urllib.robotparser.RobotFileParser()
    try:
        req = urllib.request.Request(robots_url, headers={"User-Agent": user_agent})
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            content = resp.read().decode("utf-8", "replace")
        rp.parse(content.splitlines())
    except Exception:
        return True  # no robots.txt / unreachable => allowed
    return rp.can_fetch(user_agent, url)


def web_fetch(
    url: str,
    *,
    user_agent: str = _DEFAULT_UA,
    timeout: float = 15.0,
    max_bytes: int = 2_000_000,
    respect_robots: bool = True,
) -> Dict[str, Any]:
    """Fetch ``url`` and return ``{url, status, title, text, links}``. Honours robots.txt
    unless ``respect_robots=False``. Raises ``PermissionError`` if robots disallows."""
    if urlparse(url).scheme not in ("http", "https"):
        raise ValueError(f"unsupported URL scheme: {url!r}")
    if respect_robots and not _robots_allows(url, user_agent, timeout):
        raise PermissionError(f"robots.txt disallows fetching {url}")
    req = urllib.request.Request(url, headers={"User-Agent": user_agent, "Accept": "text/html,*/*"})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        status = getattr(resp, "status", 200)
        raw = resp.read(max_bytes)
        charset = resp.headers.get_content_charset() or "utf-8"
        final_url = resp.geturl()
    html = raw.decode(charset, "replace")
    extractor = _Extractor(final_url)
    extractor.feed(html)
    return {
        "url": final_url,
        "status": status,
        "title": extractor.title,
        "text": extractor.text,
        "links": extractor.links,
    }


def register_web_tools(registry, *, respect_robots: bool = True, prefix: str = "") -> List[str]:
    """Register ``web_fetch`` (text+title) and ``web_links`` (link list) into a
    ToolRegistry. Opt-in. Returns the registered tool ids."""

    def _fetch(params: Dict[str, Any]) -> Any:
        url = params.get("url") or params.get("query") or ""
        result = web_fetch(str(url), respect_robots=respect_robots)
        # tools return a single string the brain can read back
        title = result["title"]
        head = f"{title}\n\n" if title else ""
        return head + result["text"]

    def _links(params: Dict[str, Any]) -> Any:
        url = params.get("url") or ""
        result = web_fetch(str(url), respect_robots=respect_robots)
        return "\n".join(result["links"])

    fetch_id = f"{prefix}web_fetch"
    links_id = f"{prefix}web_links"
    registry.register(fetch_id, "Fetch a web page (robots-aware) and return its title + text.", _fetch)
    registry.register(links_id, "Fetch a web page and return its outbound links, one per line.", _links)
    return [fetch_id, links_id]

File: ports/experimental/go/core/web.go
package core

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"golang.org/x/net/html"
)

// Web toolkit (Tier 4) — a robots-aware fetch + text/link extraction tool built on the
// standard library plus golang.org/x/net/html. Ships as an opt-in tool, default off: call
// RegisterWebTools(registry) to expose web_fetch / web_links to an agent. Mirrors the
// Flink web/ toolkit (Jsoup + crawler-commons robots) at a portable subset.

const defaultUserAgent = "jagentic-webfetch/0.1 (+https://github.com/jagentic)"

var skipTags = map[string]bool{"script": true, "style": true, "noscript": true, "template": true}

// FetchResult is the outcome of a web fetch.
type FetchResult struct {
	URL    string
	Status int
	Title  string
	Text   string
	Links  []string
}

// WebFetcher fetches and extracts pages, honouring robots.txt unless disabled.
type WebFetcher struct {
	UserAgent     string
	Timeout       time.Duration
	MaxBytes      int64
	RespectRobots bool
	client        *http.Client
}

// NewWebFetcher builds a fetcher with sane defaults (robots respected).
func NewWebFetcher() *WebFetcher {
	return &WebFetcher{
		UserAgent:     defaultUserAgent,
		Timeout:       15 * time.Second,
		MaxBytes:      2_000_000,
		RespectRobots: true,
		client:        &http.Client{Timeout: 15 * time.Second},
	}
}

// Fetch retrieves url and extracts its title, text and absolute links.
func (w *WebFetcher) Fetch(rawURL string) (*FetchResult, error) {
	u, err := url.Parse(rawURL)
	if err != nil || (u.Scheme != "http" && u.Scheme != "https") {
		return nil, fmt.Errorf("unsupported URL: %q", rawURL)
	}
	if w.RespectRobots {
		allowed, _ := w.robotsAllows(u)
		if !allowed {
			return nil, fmt.Errorf("robots.txt disallows fetching %s", rawURL)
		}
	}
	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", w.UserAgent)
	req.Header.Set("Accept", "text/html,*/*")
	resp, err := w.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(io.LimitReader(resp.Body, w.MaxBytes))
	if err != nil {
		return nil, err
	}
	doc, err := html.Parse(strings.NewReader(string(body)))
	if err != nil {
		return nil, err
	}
	res := &FetchResult{URL: resp.Request.URL.String(), Status: resp.StatusCode}
	extract(doc, resp.Request.URL, res, false)
	res.Text = strings.TrimSpace(res.Text)
	res.Title = strings.TrimSpace(res.Title)
	return res, nil
}

// extract walks the parse tree collecting visible text, the title, and absolute links.
func extract(n *html.Node, base *url.URL, res *FetchResult, inTitle bool) {
	if n.Type == html.ElementNode {
		tag := strings.ToLower(n.Data)
		if skipTags[tag] {
			return
		}
		if tag == "title" {
			inTitle = true
		}
		if tag == "a" {
			for _, a := range n.Attr {
				if strings.ToLower(a.Key) == "href" && a.Val != "" {
					if ref, err := base.Parse(strings.TrimSpace(a.Val)); err == nil {
						ref.Fragment = ""
						if ref.Scheme == "http" || ref.Scheme == "https" {
							res.appendLink(ref.String())
						}
					}
				}
			}
		}
	}
	if n.Type == html.TextNode {
		t := strings.TrimSpace(n.Data)
		if t != "" {
			if inTitle {
				if res.Title != "" {
					res.Title += " "
				}
				res.Title += t
			} else {
				if res.Text != "" {
					res.Text += "\n"
				}
				res.Text += t
			}
		}
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		extract(c, base, res, inTitle)
	}
}

func (r *FetchResult) appendLink(link string) {
	for _, l := range r.Links {
		if l == link {
			return
		}
	}
	r.Links = append(r.Links, link)
}

// robotsAllows checks the site's robots.txt; fail-open if unreachable, fail-closed only on
// an explicit Disallow match for our user-agent (or *).
func (w *WebFetcher) robotsAllows(u *url.URL) (bool, error) {
	robotsURL := u.Scheme + "://" + u.Host + "/robots.txt"
	req, err := http.NewRequest(http.MethodGet, robotsURL, nil)
	if err != nil {
		return true, nil
	}
	req.Header.Set("User-Agent", w.UserAgent)
	resp, err := w.client.Do(req)
	if err != nil {
		return true, nil
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return true, nil
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, 512_000))
	if err != nil {
		return true, nil
	}
	return robotsCanFetch(string(body), u.Path), nil
}

// robotsCanFetch is a minimal robots.txt evaluator: it gathers Disallow rules under the
// "*" user-agent group and denies if the path has a matching prefix. Allow rules override.
func robotsCanFetch(robots, path string) bool {
	if path == "" {
		path = "/"
	}
	var disallow, allow []string
	applies := false
	for _, line := range strings.Split(robots, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		key, val, ok := strings.Cut(line, ":")
		if !ok {
			continue
		}
		key = strings.ToLower(strings.TrimSpace(key))
		val = strings.TrimSpace(val)
		switch key {
		case "user-agent":
			applies = val == "*"
		case "disallow":
			if applies && val != "" {
				disallow = append(disallow, val)
			}
		case "allow":
			if applies && val != "" {
				allow = append(allow, val)
			}
		}
	}
	for _, a := range allow {
		if strings.HasPrefix(path, a) {
			return true
		}
	}
	for _, d := range disallow {
		if strings.HasPrefix(path, d) {
			return false
		}
	}
	return true
}

// RegisterWebTools registers web_fetch (title+text) and web_links (link list) into the
// registry. Opt-in. Returns the registered tool ids.
func RegisterWebTools(reg *ToolRegistry) []string {
	fetcher := NewWebFetcher()
	reg.Register("web_fetch", "Fetch a web page (robots-aware) and return its title + text.",
		func(params map[string]any) any {
			res, err := fetcher.Fetch(toStr(params["url"]))
			if err != nil {
				return map[string]any{"error": err.Error()}
			}
			if res.Title != "" {
				return res.Title + "\n\n" + res.Text
			}
			return res.Text
		})
	reg.Register("web_links", "Fetch a web page and return its outbound links, one per line.",
		func(params map[string]any) any {
			res, err := fetcher.Fetch(toStr(params["url"]))
			if err != nil {
				return map[string]any{"error": err.Error()}
			}
			return strings.Join(res.Links, "\n")
		})
	return []string{"web_fetch", "web_links"}
}

func toStr(v any) string {
	if v == nil {
		return ""
	}
	return fmt.Sprintf("%v", v)
}
